### DISCUSSION TITLE
iOS/Safari - continue generating text when closed - mostly solved with 1.58 (run in background)

### DISCUSSION THREAD

**Discussion by FrostyMisa** [LostRuins/koboldcpp#633] [Q&A]
Do you think it will be possible (or maybe it's and I don't find how) to continue the text generation even I close Safari under iOS and Koboldcpp will continue generate the text, after it will save this text locally on the machine where is the server, so I can only open Safari, refresh the website and the generated answer will appear here? 

Because now when I close Safari, the answer is lost when I don't want to wait in the web interface. Thanks, love Koboldcpp!

**Comment by LostRuins** [comment]
You should be able to minimize the browser while the generation is ongoing. However, exiting the browser would cause the response to be lost, though it should still appear in the terminal console. 

What you're requesting may be possible with a third party client when using the API directly though, you just need a separate program/script to request and save the generation instead of using your browser.

**Reply by FrostyMisa** [reply]
Yeah, on PC or the server machine is not problem leave browser open. But on mobile device it's not so convenient wait one minute with opened browser for answer. I'm running Koboldcpp on mini pc, low power, running on CPU, is good, but slower. 

And because the message stay in log I hope there will be some way to call it from there when I open the browser and it doesn't get lost. And Koboldcpp already support internal web interface(I think it's the Kobold lite?), so something like it will read from some database, where it will be saved locally when the client cannot be reached?

And Safari for iOS also support delivery by push notification, so this will be the best solution! Write message, close Safari, use another app, after it's done, get notification. Koboldcpp will be running, so I think it can send the push as well after generating. 

https://developer.apple.com/documentation/usernotifications/sending_web_push_notifications_in_web_apps_and_browsers

Koboldcpp is great, this small thing will improve usability on iOS devices a lot, mostly running on slower CPU pc. Thanks for your hard work guys!

**Reply by LostRuins** [reply]
The server does not store the prompt after generation is completed. It's just output to the terminal, but the context is managed by the browser client. That's why multiple devices can share the same server with different stories.

**Reply by FrostyMisa** [reply]
Yeah, I understand this. And do you think it's possible to add the Safari push notification to the Koboldcpp? Or for different browser for iOS(chrome), if it will be more easy to implement.  Thanks for your time.

**Reply by FrostyMisa** [reply]
I tested the latest version of Koboldcpp and the notification for chrome on PC in windows works great. After first message, it will ask if I want receive notification.
But sadly, on iOS, I tested in Safari, chrome and Firefox and it doesn’t ask me for receive notification.  So there must be some other way for iOS devices.

**Reply by LostRuins** [reply]
Hmm... did you enable notifications in your settings? I don't use iOS but maybe there are some security settings that toggle browser notifications.

**Reply by LostRuins** [reply]
I tried firefox on windows and it worked fine.

**Reply by FrostyMisa** [reply]
Yeah, on windows works great! 👍 But on iOS it doesn't ask for them. Maybe the implementation will be little bit different for iOS, because all the browsers are force to use Apple WebKit. 

I think how to implement it is in the link above I send before. I will send it again. 

https://developer.apple.com/documentation/usernotifications/sending_web_push_notifications_in_web_apps_and_browsers

They write "You don’t need to join the Apple Developer Program to send web push notifications."

https://developer.mozilla.org/en-US/docs/Web/API/Push_API

Only Safari for iOS in the list have some special implementation. 

You can ask someone you know with iOS to test it. This website push works for me, so it's not problem with settings on my phone. 

https://pushalert.co/push-notifications-for-safari

Sometimes I'm angry some do some things differently, like in this case. I hope you will find some time to implement it also for iOS. 

I appreciate your time you already invested in this❤️

**Reply by LostRuins** [reply]
For now, you can try enabling "Beep on Done" which will sound a tone when generation is completed. That may work in place of notifications.

**Reply by FrostyMisa** [reply]
I tried it and even beep doesn't work. I checked if I don't have silent mode.

I thinking if the problem is not SSL certificate, because I don't use it locally, but it's not, because I tried it on
https://lite.koboldai.net/
and even here it doesn't work (push and beep). 

Thank you for your time. I really appreciate it!

**Reply by FrostyMisa** [reply]
I only want to say thanks for the last update! Even the notification doesn't work on iOS (works on desktop PC with chrome)(I think it will be more complicated to make it work for Apple ecosystem), the latest version 1.58 with the settings run in background works even on iOS!

I test it, when I turn it on (run in background) and type something inside Kobold lite, I can safely close the browser and use other apps, even lock the iPhone screen and when I come back, the response waiting for me here! And in command line, where KoboldCPP running, it's no longer message interrupted by user close the browser!

Now I don't need to watch opened browser on my phone to get response, when I'm running KCPP on slow machine. 

So one more time thank you very much! 🙏❤️

**Reply by LostRuins** [reply]
You're welcome! Hope its useful for you
